lib/web/apiserver_test.go

```diff
@@ -2012,6 +2013,141 @@ func TestCreatePrivilegeToken(t *testing.T) {
 	require.NotEmpty(t, privilegeToken)
 }
 
+func TestAddMFADevice(t *testing.T) {
+	t.Parallel()
+	ctx := context.Background()
+	env := newWebPack(t, 1)
+	proxy := env.proxies[0]
+	pack := proxy.authPack(t, "foo@example.com")
+
+	// Enable second factor.
+	ap, err := types.NewAuthPreference(types.AuthPreferenceSpecV2{
+		Type:         constants.Local,
+		SecondFactor: constants.SecondFactorOptional,
+		U2F: &types.U2F{
+			AppID:  "https://localhost",
+			Facets: []string{"https://localhost"},
+		},
+	})
+	require.NoError(t, err)
+	err = env.server.Auth().SetAuthPreference(ctx, ap)
+	require.NoError(t, err)
+
+	// Get a totp code to re-auth.
+	totpCode, err := totp.GenerateCode(pack.otpSecret, env.clock.Now().Add(30*time.Second))
+	require.NoError(t, err)
+
+	// Obtain a privilege token.
+	endpoint := pack.clt.Endpoint("webapi", "users", "privilege", "token")
+	re, err := pack.clt.PostJSON(ctx, endpoint, &privilegeTokenRequest{
+		SecondFactorToken: totpCode,
+	})
+	require.NoError(t, err)
+	var privilegeToken string
+	require.NoError(t, json.Unmarshal(re.Bytes(), &privilegeToken))
+
+	tests := []struct {
+		name            string
+		deviceName      string
+		getTOTPCode     func() string
+		getU2FResp      func() *u2f.RegisterChallengeResponse
+		getWebauthnResp func() *wanlib.CredentialCreationResponse
+	}{
+		{
+			name:       "new TOTP device",
+			deviceName: "new-totp",
+			getTOTPCode: func() string {
+				// Create totp secrets.
+				res, err := env.server.Auth().CreateRegisterChallenge(ctx, &apiProto.CreateRegisterChallengeRequest{
+					TokenID:    privilegeToken,
+					DeviceType: apiProto.DeviceType_DEVICE_TYPE_TOTP,
+				})
+				require.NoError(t, err)
+
+				// Get the new totp code.
+				newTOTPCode, err := totp.GenerateCode(res.GetTOTP().GetSecret(), env.clock.Now())
```

Comment:
Could we move the device creation and signing to TestDevices? Same for other tests.